koswat/dike_reinforcements/reinforcement_profile/berm_calculator/berm_calculator_factory.py

```diff
@@ -0,0 +1,208 @@
+from collections import defaultdict
+
+from koswat.configuration.settings.koswat_scenario import KoswatScenario
+from koswat.configuration.settings.reinforcements.koswat_reinforcement_settings import (
+    KoswatReinforcementSettings,
+)
+from koswat.dike.koswat_input_profile_protocol import KoswatInputProfileProtocol
+from koswat.dike_reinforcements.input_profile.input_profile_enum import InputProfileEnum
+from koswat.dike_reinforcements.reinforcement_profile.berm_calculator import (
+    BermCalculatorProtocol,
+    DefaultBermCalculator,
+    KeepBermCalculator,
+    NoBermCalculator,
+    PipingBermCalculator,
+    StabilityBermCalculator,
+)
+
+
+class BermCalculatorFactory:
+    """
+    Factory to create the correct berm calculator based on the base profile,
+    reinforcement settings, scenario and profile type.
+    """
+
+    base_data: KoswatInputProfileProtocol
+    reinforced_data: KoswatInputProfileProtocol
+    reinforcement_settings: KoswatReinforcementSettings
+    scenario: KoswatScenario
+    _dikebase_piping_old: float
+    _dikebase_piping_new_dict: dict[InputProfileEnum, float] = defaultdict(float)
+    _dikebase_height_new: float
+    _dikebase_stability_new: float
+    _berm_old_is_stability: bool
+    _berm_factor_old: float
+    _dike_height_new: float
+    _reinforcement_type: InputProfileEnum
+
+    def __init__(
+        self,
+        base_data: KoswatInputProfileProtocol,
+        reinforced_data: KoswatInputProfileProtocol,
+        reinforcement_settings: KoswatReinforcementSettings,
+        scenario: KoswatScenario,
+    ) -> None:
```

Comment:
If I see the required arguments and how this factory is being used i believe we are not being consistent. In principle the factory should, given a provided input, return a subtype of its domain (`BermCalculatorProtocol`). Most of the times we use a static method or a class without constructor's parameters to directly get the desired functionality. Based on this I would like to see the following for consistency:

1. Class (or dataclass) for the input factors ( `_calculate_factors`), which can be initialized with `base_data`, `reinforced_data`, `reinforcement_settings` and `scenario`. Only those relevant parameters for the factory should be exposed as public. It should also be extended with the profile type.
2. Factory does not have properties nor parameters in its `__init__`. Instead the `get_berm_calculator` consumes the previously defined object to return the corresponding concrete class of `BermCalculatorProtocol`.

Why this? When we declare parameters in the constructor we are "showing" that the factory will have a base state based on them, so that afterwards you can get different results with the `get_berm_calculators` as it requires a specific `InputProfileEnum` argument. This is interesting if we would have a list of objects that need to be mapped to different calculators however always based on the aforementioned arguments. However, as earlier mentioned, when we see how this factory is being used we can notice that its base state is not being "reused" for another mapping but instead being self-consumed. Therefore this approach does not have place in the current code base.

Reply:
Introduced dataclass `BermCalculatedFactors`, with a classmethod to construct the object from the input, that is passed as input to the factory, which now only has static methods.